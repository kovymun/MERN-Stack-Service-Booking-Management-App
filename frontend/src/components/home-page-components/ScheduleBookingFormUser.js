/**TAC SERVICE BOOKING APP SCHEDULE BOOKING FORM REACT COMPONENT FILE - NORMAL USER (SERVICE BOOKING AGENT)**/
/**KINDLY REFER TO THE END OF THIS DOCUMENT FOR ALL REFERENCES**/

import { v4 as uuidv4 } from "uuid";
import { Button, Form, Col, Row, FormSelect } from "react-bootstrap";

/*
TAC Service Booking App: Schedule new service booking form component:
This component affords the service booking agents (normal user) the ability to schedule new customer service bookings. The information fields to be entered by 
the user are the following:
* The Customer Name.
* The Customer Contact Details.
* The Vehicle Make, Model and Registration number.
* Service booking date and time.
* Service to be delivered to the customer.
*/

/* Array containing objects representing each service type offered by Tyler's Auto Clinic. */
const services = [
  { serviceType: "Car: Battery Testing" },
  {
    serviceType: "Car: Full Details Vehicle Safety Assessment",
  },
  { serviceType: "Wheels: Wheel Balancing and Alignment" },
  {
    serviceType: "Wheels: Tyre Puncture Assessment and Repairs",
  },
  { serviceType: "Engine: Oil and Filter Replacement" },
];

export default function ScheduleBookingFormUser({
  handleAddBooking,
  handleServiceType,
  handleDate,
  customerValue,
  numValue,
  vehMake,
  vehModel,
  vehReg,
  addBooking,
}) {
  return (
    <div className="schedule-booking-form">
      <h5 className="schedule-booking-title">Schedule Booking</h5>
      <Form onSubmit={addBooking}>
        {/* Form input for the user to enter the name of the customer */}
        <Form.Group as={Row}>
          <h6 className="customer">Customer:</h6>
          <Col lg="12">
            <Form.Control
              name="customer"
              value={customerValue}
              onChange={handleAddBooking}
              className="mb-2"
              placeholder="Enter Customer Name e.g. Calvin Harris"
            />
          </Col>
        </Form.Group>

        {/* Form input for the user to enter the contact number of the customer */}
        <Form.Group as={Row}>
          <h6 className="contact-number">Customer Contact Number:</h6>
          <Col lg="12">
            <Form.Control
              name="customerContactNumber"
              value={numValue}
              onChange={handleAddBooking}
              className="mb-2"
              placeholder="Enter Customer Contact No. e.g. 083XXXXXXX"
            />
          </Col>
        </Form.Group>

        {/* Form input for the user to enter the make/manufacturer of the vehicle */}
        <Form.Group as={Row}>
          <h6 className="make">Vehicle Make:</h6>
          <Col lg="12">
            <Form.Control
              name="vehicleMake"
              value={vehMake}
              onChange={handleAddBooking}
              className="mb-2"
              placeholder="Enter Vehicle Make e.g. Ford Mustang"
            />
          </Col>
        </Form.Group>

        {/* Form input for the user to enter the model year of the vehicle */}
        <Form.Group as={Row}>
          <h6 className="model">Vehicle Model:</h6>
          <Col lg="12">
            <Form.Control
              name="vehicleModel"
              value={vehModel}
              onChange={handleAddBooking}
              className="mb-2"
              placeholder="Enter Vehicle Model e.g. 2009"
              maxLength="4"
            />
          </Col>
        </Form.Group>

        {/* Form input for the user to enter the vehicle registration number */}
        <Form.Group as={Row}>
          <h6 className="registration">Vehicle Reg:</h6>
          <Col lg="12">
            <Form.Control
              name="vehicleReg"
              value={vehReg}
              onChange={handleAddBooking}
              className="mb-2"
              placeholder="Enter Vehicle Registration e.g. GP12375"
              maxLength="10"
            />
          </Col>
        </Form.Group>

        {/* date-time picker form to select the service booking date and time */}
        <Form.Group as={Row}>
          <h6 className="date">Date:</h6>
          <Col lg="12">
            <Form.Control
              name="date"
              type="datetime-local"
              onChange={handleDate}
              className="mb-2"
            />
          </Col>
        </Form.Group>

        {/* Form select options for the user to select the service to be delivered to the customer */}
        <Form.Group as={Row}>
          <h6 className="service">Service:</h6>
          <Col lg="12">
            <FormSelect onChange={handleServiceType}>
              {/* Map through services array to create a list of options to select from */}
              {services.map((service) => (
                <option key={uuidv4()} value={service.serviceType}>
                  {service.serviceType}
                </option>
              ))}
            </FormSelect>
          </Col>
        </Form.Group>
        <Button className="mt-3" variant="light" type="submit">
          SCHEDULE BOOKING
        </Button>
      </Form>
    </div>
  );
}

/**REFERENCES**/
/*
React-Bootstrap Components:
Utilized the React-Bootstrap Button, Form, Col, Row components as an interface, for the service booking agent (normal user) to enter the relevant information 
to schedule the booking.
Referenced the React-Bootstrap documentation website for this information. 
https://react-bootstrap.github.io/

maxLength attribute:
Researched and utilized the maxLength HTML attribute, to limit the number of characters the user can enter for a particular field.
Referenced the w3schools website for this information.
https://www.w3schools.com/tags/att_input_maxlength.asp
*/
