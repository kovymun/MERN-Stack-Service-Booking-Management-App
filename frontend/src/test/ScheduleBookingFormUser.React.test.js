/**TAC SERVICE BOOKING APP SCHEDULE BOOKING COMPONENT JEST SNAPSHOT TEST FILE**/
/**KINDLY REFER TO THE END OF THIS DOCUMENT FOR ALL REFERENCES**/

import React from "react";
import renderer from "react-test-renderer";
import ScheduleBookingFormUser from "../components/home-page-components/ScheduleBookingFormUser";

it("ScheduleBookingFormUser component renders correctly", () => {
  const tree = renderer.create(<ScheduleBookingFormUser />).toJSON();
  expect(tree).toMatchSnapshot();
});

/**REFERENCES**/
/*
JEST Snapshot testing:
Referenced the JEST testing framework website for information on how to perform snapshot testing on a component.
https://jestjs.io/docs/snapshot-testing
*/
