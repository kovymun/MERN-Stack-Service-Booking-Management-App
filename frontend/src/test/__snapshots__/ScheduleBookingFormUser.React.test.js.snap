// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleBookingFormUser component renders correctly 1`] = `
<div
  className="schedule-booking-form"
>
  <h5
    className="schedule-booking-title"
  >
    Schedule Booking
  </h5>
  <form
    className=""
  >
    <div
      className="row"
    >
      <h6
        className="customer"
      >
        Customer:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          name="customer"
          placeholder="Enter Customer Name e.g. Calvin Harris"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="contact-number"
      >
        Customer Contact Number:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          name="customerContactNumber"
          placeholder="Enter Customer Contact No. e.g. 083XXXXXXX"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="make"
      >
        Vehicle Make:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          name="vehicleMake"
          placeholder="Enter Vehicle Make e.g. Ford Mustang"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="model"
      >
        Vehicle Model:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          maxLength="4"
          name="vehicleModel"
          placeholder="Enter Vehicle Model e.g. 2009"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="registration"
      >
        Vehicle Reg:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          maxLength="10"
          name="vehicleReg"
          placeholder="Enter Vehicle Registration e.g. GP12375"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="date"
      >
        Date:
      </h6>
      <div
        className="col-lg-12"
      >
        <input
          className="mb-2 form-control"
          name="date"
          type="datetime-local"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <h6
        className="service"
      >
        Service:
      </h6>
      <div
        className="col-lg-12"
      >
        <select
          className="form-select"
        >
          <option
            value="Car: Battery Testing"
          >
            Car: Battery Testing
          </option>
          <option
            value="Car: Full Details Vehicle Safety Assessment"
          >
            Car: Full Details Vehicle Safety Assessment
          </option>
          <option
            value="Wheels: Wheel Balancing and Alignment"
          >
            Wheels: Wheel Balancing and Alignment
          </option>
          <option
            value="Wheels: Tyre Puncture Assessment and Repairs"
          >
            Wheels: Tyre Puncture Assessment and Repairs
          </option>
          <option
            value="Engine: Oil and Filter Replacement"
          >
            Engine: Oil and Filter Replacement
          </option>
        </select>
      </div>
    </div>
    <button
      className="mt-3 btn btn-light"
      disabled={false}
      type="submit"
    >
      SCHEDULE BOOKING
    </button>
  </form>
</div>
`;
